ObjectOrientedProgramming/Memory Ownership/move semantics/task.md

```diff
@@ -0,0 +1,54 @@
+Move semantics is a feature of C++ that allows you to efficiently transfer ownership of an object from one variable to another without copying the object. This can be useful for improving performance and avoiding unnecessary memory allocations.
+
+#### Move Semantics Overview
+- Efficient Resource Transfer. Move semantics allows the transfer of ownership from one object to another without the need for deep copying. This is especially beneficial for large or resource-intensive objects. 
+- Rvalue References. Move semantics utilize rvalue references (`&&`), which represent temporary objects or objects that are about to be destroyed. Rvalue references to facilitate the efficient transfer of ownership without unnecessary overhead.
+- `std::move` Function. The `std::move` function is used to convert a lvalue (an object with a name) into a rvalue reference. This is a key tool for explicitly indicating that ownership can be moved.
+
+#### Move Semantics with `std::unique_ptr`
+`std::unique_ptr` supports move semantics through its `std::move()` member function. The `std::move()` function transfers ownership of the object from the source `std::unique_ptr` to the destination unique_ptr. The source `std::unique_ptr` is then left in a null state.
+
+Here is an example of how to use the `std::move()` function to transfer ownership of a `std::unique_ptr` object:
+
+```cpp
+#include <memory>
+
+class MyClass {
+public:
+    MyClass() {
+        std::cout << "MyClass Constructor\n";
+    }
+
+    ~MyClass() {
+        std::cout << "MyClass Destructor\n";
+    }
+};
+
+int main() {
+    // Create two unique_ptr objects.
+    std::unique_ptr<MyClass> p1 = std::make_unique<MyClass>();
+    std::unique_ptr<MyClass> p2;
+
+    // Transfer ownership of p1 to p2.
+    p2 = std::move(p1);
+
+    // p1 is now in a null state.
+    assert(p1 == nullptr);
+
+    // p2 now owns the MyClass object.
+    assert(p2 != nullptr);
+
+    return 0;
+}
+```
+
+Move semantics can be useful for a variety of tasks, such as:
+
+- Passing a `std::unique_ptr` object to a function that takes ownership of the object.
+- Returning a `std::unique_ptr` object from a function.
+- Moving a `std::unique_ptr` object from one container to another.
+
+Additional links:
+- [std::move](https://en.cppreference.com/w/cpp/utility/move)
+- [move semantics](https://en.cppreference.com/w/cpp/language/move_constructor)
+- [rvalue references](https://en.cppreference.com/w/cpp/language/rvalue_reference)
```

Comment:
In previous lessons, links to cppreference were given directly in the text (not at the end of task as a list).

We should try to maintain consistency in presentation throughout the course. 
We can discuss what style is better on a meeting.